benchmarks/read.js

```diff
@@ -0,0 +1,266 @@
+      }
+
+      if (writes >= options.n) {
+        return callback()
+      }
+
+      const batchSize = Math.min(1000, options.n - writes)
+      const ops = new Array(batchSize)
+
+      for (let i = 0; i < batchSize; i++) {
+        const key = writeGenerator.key(writes++)
+        const value = writeGenerator.value()
+
+        // TODO: what if db has encodings?
+        ops[i] = { type: 'put', key, value }
+      }
+
+      // Wait for last batch to be flushed to disk (if supported)
+      const isLastBatch = writes >= options.n
+      const batchOptions = { sync: isLastBatch }
+
+      db.batch(ops, batchOptions, loop)
+    }
+  }
+
+  function compact (db, callback) {
+    // Flip start & end because writeGenerator is in reverse
+    const start = writeGenerator.key(options.n - 1)
+    const end = writeGenerator.key(0)
+
+    lcompact(db, start, end, delayed(function (err) {
+      if (err) return callback(err)
+
+      ldu(db, function (err, size) {
+        if (err) return callback(err)
+        if (size) console.log('Size', bytes.format(size))
+        callback()
+      })
+    }, 500))
+  }
+
+  function start (db) {
+    const startTime = Date.now()
+
+    // TODO: what if db has encodings?
+    const getOptions = Object.assign({}, isOptions(options.get) ? options.get : {}, {
+      asBuffer: !!options.valueAsBuffer
+    })
+
+    // TODO: what if db has encodings?
+    const iteratorOpts = Object.assign({}, isOptions(options.iterator) ? options.iterator : {}, {
+      keyAsBuffer: !!options.keyAsBuffer,
+      valueAsBuffer: !!options.valueAsBuffer
+    })
+
+    let inProgress = 0
+    let totalReads = 0
+    let totalBytes = 0
+    let timesAccum = 0
+    let lastKey
+    let elapsed
+    let iterator
+
+    function report () {
+      if (inProgress) return
+
+      console.log(
+        'Read', options.n, 'entries in',
+        Math.floor((Date.now() - startTime) / 1e3) + 's,',
+        (Math.floor((totalBytes / 1048576) * 100) / 100) + 'MB'
+      )
+
+      // TODO: end iterator and/or read stream
+      stream.end()
+
+      db.close(function (err) {
+        if (err) throw err
+      })
+    }
+
+    function read () {
+      if (inProgress >= options.concurrency) return
+      if (totalReads >= options.n) return report()
+
+      inProgress++
+
+      if (totalReads % progressWindow === 0) {
+        console.log('' + inProgress, totalReads,
+          Math.round(totalReads / options.n * 100) + '%')
+      }
+
+      if (totalReads % window === 0) {
+        elapsed = Date.now() - startTime
+        stream.write(
+          elapsed +
+          ',' + totalReads +
+          ',' + totalBytes +
+          ',' + (timesAccum / window / 1e6).toFixed(3) +
+          ',' + ((totalBytes / 1048576) / (elapsed / 1e3)).toFixed(3) +
+          '\n')
+        timesAccum = 0
+      }
+
+      let start
+
+      if (options.iterator) {
+        if (options.concurrency > 1) {
+          // TODO: create pool of iterators?
```

Comment:
@peakji WDYT would be the most useful behavior? If you want to benchmark concurrent and/or multiple successive iterators, should they all read the same range, a random range, or sequential ranges (e.g. the first iterator reads 0-10, the second one reads 10-20)?

Reply:
It's kinda weird to use multiple iterators to read the same range, I can't come up with a real work use case at the moment.

Since we are talking about benchmarks, I would say random (or *distant*) range might be the better choice: Sequential range reading may cheat the benchmark, given the fact that leveldb (and most of the other backends) is reading with blocks, successive iterators can often read from a block that is already in some sort of cache.

Reply:
I knew I asked the right person :)

How about simply starting at a random key? Given enough keys, a small `limit` and a uniform random distribution, that should result in enough distance between the iterators.

Reply:
👍 Random keys should work.